File: LeetCode/Study Plan/Practice Problems/Medium/Koko Eating Bananas/Go/koko.go
package main

import (
	"fmt"
	"math"
)

func minEatingSpeed(piles []int, h int) int {
	maxPile := math.MinInt64
	size := len(piles)

	for i := 0; i < size; i++ {
		maxPile = max(maxPile, piles[i])
	}

	l := 1
	r := maxPile

	for l < r {
		mid := l + (r-l)/2
		if isPossible(h, mid, piles) {
			r = mid
		} else {
			l = mid + 1
		}
	}

	return r
}

func isPossible(h, banana int, piles []int) bool {
	hrs := 0
	size := len(piles)

	for i := 0; i < size; i++ {
		hrs += int(math.Ceil(float64(piles[i]) / float64(banana)))
	}

	return hrs <= h
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(minEatingSpeed([]int{3, 6, 7, 11}, 8))
}

File: LeetCode/Medium/Koko Eating Bananas/Java/koko.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int maxPile = 0;

        for (int pile : piles) {
            maxPile = Math.max(maxPile, pile);
        }

        int l = 1;
        int r = maxPile;

        while(l < r) {
            int mid = l + (r-l)/2;
            if (isPossible(mid, piles, h)) {
                r = mid;
            }else {
                l = mid+1;
            }
        }

        return r;
    }

    public boolean isPossible(int banana, int[] piles, int h) {
        double hrs = 0;
        for (int pile : piles) {
            hrs += Math.ceil((double)pile/banana);
        }

        return hrs<=h;
    }
}
